core/trino-main/src/main/java/io/trino/sql/relational/SqlToRowExpressionTranslator.java

```diff
@@ -702,7 +702,7 @@ protected RowExpression visitSubscriptExpression(SubscriptExpression node, Void
 
             if (getType(node.getBase()) instanceof RowType) {
                 long value = (Long) ((ConstantExpression) index).getValue();
-                return new SpecialForm(DEREFERENCE, getType(node), base, constant((int) value - 1, INTEGER));
+                return new SpecialForm(DEREFERENCE, getType(node), base, constant(value - 1, INTEGER));
```

Comment:
Unrelated change?

Reply:
Oh, this is *the* fix. It's not at all obvious. Can you add an explanation in the commit message?

Reply:
it's related

Reply:
Yes, I know. That's what I said above. Please add an explanation in the commit message, as it's not immediately obvious why this fixes the bug.

Reply:
> Yes, I know. That's what I said above.

github somehow showed me the first comment, but not the second (i should have refreshed the page before replying)

Reply:
> Please add an explanation in the commit message, as it's not immediately obvious why this fixes the bug.

rephrased commit message as

```
    Fix ConstantExpression serialization

    When `ConstantExpression` was created with a Trino type and a value
    that didn't match this type's java type (for example Trino INTEGER with
    value being Integer, instead of Long), the `ConstantExpression` would
    fail to serialize to JSON. Such failure happens during sending task
    status updates to workers and is currently logged and ignored, leading
    to query hang. The problem could be triggered with SQL routines, which
    utilize `RowExpression` (including `ConstantExpression`) serialization.

    Thus, this fixes execution of SQL routines involving Row field
    dereference.
```